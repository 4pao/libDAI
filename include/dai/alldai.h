/*  Copyright (C) 2006-2008  Joris Mooij  [joris dot mooij at tuebingen dot mpg dot de]
    Radboud University Nijmegen, The Netherlands /
    Max Planck Institute for Biological Cybernetics, Germany

    This file is part of libDAI.

    libDAI is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    libDAI is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with libDAI; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/


/// \file
/// \brief Main libDAI header file
/// \todo Improve documentation


#ifndef __defined_libdai_alldai_h
#define __defined_libdai_alldai_h


#include <string>
#include <dai/daialg.h>
#include <dai/properties.h>
#include <dai/exactinf.h>
#ifdef DAI_WITH_BP
    #include <dai/bp.h>
#endif
#ifdef DAI_WITH_MF
    #include <dai/mf.h>
#endif
#ifdef DAI_WITH_HAK
    #include <dai/hak.h>
#endif
#ifdef DAI_WITH_LC
    #include <dai/lc.h>
#endif
#ifdef DAI_WITH_TREEEP
    #include <dai/treeep.h>
#endif
#ifdef DAI_WITH_JTREE
    #include <dai/jtree.h>
#endif
#ifdef DAI_WITH_MR
    #include <dai/mr.h>
#endif


/// Namespace for libDAI
namespace dai {


/// Constructs a new approximate inference algorithm.
/** \param name The name of the approximate inference algorithm (should be one of the names in DAINames).
 *  \param fg The FactorGraph that the algorithm should be applied to.
 *  \param opts A PropertySet specifying the options for the algorithm.
 *  \return Returns a pointer to the new InfAlg object; it is the responsibility of the caller to delete it later.
 */
InfAlg *newInfAlg( const std::string &name, const FactorGraph &fg, const PropertySet &opts );


/// Contains the names of all approximate inference algorithms compiled into libDAI.
static const char* DAINames[] = {
    ExactInf::Name,
#ifdef DAI_WITH_BP
    BP::Name, 
#endif
#ifdef DAI_WITH_MF
    MF::Name,
#endif
#ifdef DAI_WITH_HAK
    HAK::Name,
#endif
#ifdef DAI_WITH_LC
    LC::Name,
#endif
#ifdef DAI_WITH_TREEEP
    TreeEP::Name,
#endif
#ifdef DAI_WITH_JTREE
    JTree::Name,
#endif
#ifdef DAI_WITH_MR
    MR::Name,
#endif
    ""
};


} // end of namespace dai


#endif
