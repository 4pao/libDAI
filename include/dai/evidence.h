/*
  Copyright 2009 Charles Vaske <cvaske@soe.ucsc.edu>
  University of California Santa Cruz

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.
  
  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
  
  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#ifndef __defined_libdai_evidence_h
#define __defined_libdai_evidence_h

#include <istream>

#include <dai/daialg.h>

namespace dai {

/// Store joint observations on a graphical model.
class SampleData {
private:
  std::string _name;
  std::map<Var, size_t> _obs;
public:
  /// Construct an empty object
  SampleData() : _name(), _obs() {}
  /// Set the name of the sample
  void name(const std::string& name) { _name = name; }
  /// Get the name of the sample
  const std::string& name() const { return _name; }
  /// Read from the observation map
  const std::map<Var, size_t>& observations() const { return _obs; }
  /// Add an observation
  void addObservation(Var node, size_t setting);
  /// Add evidence by clamping variables to observed values.
  void applyEvidence(InfAlg& alg) const;
};

/// Store observations from a graphical model.
class Evidence {
private:
  std::map< std::string, SampleData > _samples;
public:
  /// Start with empty obects, then fill with calls to addEvidenceTabFile()
 Evidence() : _samples() {}
  
  /** Read in tab-data from a stream. Each line contains one sample, and
   * the first line is a header line with names. The first column contains
   * names for each of the samples.
   */
  void addEvidenceTabFile(std::istream& is,
			  std::map< std::string, Var >& varMap);

  /** Read in tab-data from a stream. Each line contains one sample,
   * and the first line is a header line with variable IDs. The first
   * column contains names for each of the samples.
   */
  void addEvidenceTabFile(std::istream& is, FactorGraph& fg);
  
  /// Total number of samples in this evidence file
  size_t nrSamples() const { return _samples.size(); }

  /// @name iterator interface
  //@{
  typedef std::map< std::string, SampleData >::iterator iterator;
  typedef std::map< std::string, SampleData >::const_iterator const_iterator;
  iterator begin() { return _samples.begin(); }
  const_iterator begin() const { return _samples.begin(); }
  iterator end() { return _samples.end(); }
  const_iterator end() const { return _samples.end(); }
  //@}

};
  
}

#endif
