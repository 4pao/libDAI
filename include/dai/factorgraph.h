/*  Copyright (C) 2006-2008  Joris Mooij  [joris dot mooij at tuebingen dot mpg dot de]
    Radboud University Nijmegen, The Netherlands /
    Max Planck Institute for Biological Cybernetics, Germany

    This file is part of libDAI.

    libDAI is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    libDAI is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with libDAI; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/


/// \file
/// \brief Defines the FactorGraph class
/// \todo Improve documentation


#ifndef __defined_libdai_factorgraph_h
#define __defined_libdai_factorgraph_h


#include <iostream>
#include <map>
#include <dai/bipgraph.h>
#include <dai/factor.h>


namespace dai {


/// Represents a factor graph.
/** Both Bayesian Networks and Markov random fields can be represented in a 
 *  unifying representation, called <em>factor graph</em> [\ref KFL01], 
 *  implemented in libDAI by the FactorGraph class.
 *  
 *  Consider a probability distribution over \f$N\f$ discrete random variables 
 *  \f$x_0,x_1,\dots,x_N\f$ that factorizes as a product of factors, each of
 *  which depends on some subset of the variables:
 *  \f[
 *    P(x_0,x_1,\dots,x_N) = \frac{1}{Z} \prod_{I=0}^M f_I(x_I), \qquad
 *    Z = \sum_{x_0}\dots\sum_{x_N} \prod_{I=0}^M f_I(X_I).
 *  \f]
 *  Each factor \f$f_I\f$ is a function from an associated subset
 *  of variables \f$X_I \subset \{x_0,x_1,\dots,x_N\}\f$ to the nonnegative
 *  real numbers.
 * 
 *  For a Bayesian network, each factor corresponds to a (conditional) 
 *  probability table, whereas for a Markov random field, each factor 
 *  corresponds to a maximal clique of the undirected graph.
 *
 *  Factor graphs explicitly express the factorization structure of the
 *  corresponding probability distribution.
 *
 *  \todo Alternative implementation of undo factor changes: the only things that have to be
 *  undone currently are setting a factor to 1 and setting a factor to a Kronecker delta. This
 *  could also be implemented in the TFactor itself, which could maintain its state 
 *  (ones/delta/full) and act accordingly.
 */ 
class FactorGraph {
    public:
        /// Stores the neighborhood structure
        BipartiteGraph                    G;

        /// Shorthand for BipartiteGraph::Neighbor
        typedef BipartiteGraph::Neighbor  Neighbor;

        /// Shorthand for BipartiteGraph::Neighbors
        typedef BipartiteGraph::Neighbors Neighbors;

        /// Shorthand for BipartiteGraph::Edge
        typedef BipartiteGraph::Edge      Edge;
        
        /// Iterator over factors
        typedef std::vector<Factor>::iterator iterator;
        
        /// Const iterator over factors
        typedef std::vector<Factor>::const_iterator const_iterator;
        

    private:
        std::vector<Var>         _vars;
        std::vector<Factor>      _factors;
        std::map<size_t,Factor>  _backup;

    public:
        /// Default constructor
        FactorGraph() : G(), _vars(), _factors(), _backup() {}

        /// Constructs a FactorGraph from a vector of factors
        FactorGraph(const std::vector<Factor> &P);

        /// Constructs a FactorGraph from given factor and variable iterators
        /** \tparam FactorInputIterator Iterator with value_type Factor
         *  \tparam VarInputIterator Iterator with value_type Var
         *  \pre Assumes that the set of variables in [var_begin,var_end) is the union of the variables in the factors in [fact_begin, fact_end)
         */
        template<typename FactorInputIterator, typename VarInputIterator>
        FactorGraph(FactorInputIterator fact_begin, FactorInputIterator fact_end, VarInputIterator var_begin, VarInputIterator var_end, size_t nr_fact_hint = 0, size_t nr_var_hint = 0 );

        /// Destructor        
        virtual ~FactorGraph() {}

        /// Clone *this (virtual copy constructor)
        virtual FactorGraph* clone() const { return new FactorGraph(); }

        /// Create (virtual default constructor)
        virtual FactorGraph* create() const { return new FactorGraph(*this); }

        /// Returns const reference to i'th variable
        const Var & var(size_t i) const { return _vars[i]; }
        /// Returns const reference to all factors
        const std::vector<Var> & vars() const { return _vars; }
        /// Returns reference to I'th factor
        Factor & factor(size_t I) { return _factors[I]; }
        /// Returns const reference to I'th factor
        const Factor & factor(size_t I) const { return _factors[I]; }
        /// Returns const reference to all factors
        const std::vector<Factor> & factors() const { return _factors; }
        /// Returns iterator pointing to first factor
        iterator begin() { return _factors.begin(); }
        /// Returns const iterator pointing to first factor
        const_iterator begin() const { return _factors.begin(); }
        /// Returns iterator pointing beyond last factor
        iterator end() { return _factors.end(); }
        /// Returns const iterator pointing beyond last factor
        const_iterator end() const { return _factors.end(); }

        /// Returns number of variables
        size_t nrVars() const { return vars().size(); }
        /// Returns number of factors
        size_t nrFactors() const { return factors().size(); }
        /// Calculates number of edges
        size_t nrEdges() const { return G.nrEdges(); }

        /// Provides read access to neighbors of variable
        const Neighbors & nbV( size_t i ) const { return G.nb1(i); }
        /// Provides full access to neighbors of variable
        Neighbors & nbV( size_t i ) { return G.nb1(i); }
        /// Provides read access to neighbors of factor
        const Neighbors & nbF( size_t I ) const { return G.nb2(I); }
        /// Provides full access to neighbors of factor
        Neighbors & nbF( size_t I ) { return G.nb2(I); }
        /// Provides read access to neighbor of variable
        const Neighbor & nbV( size_t i, size_t _I ) const { return G.nb1(i)[_I]; }
        /// Provides full access to neighbor of variable
        Neighbor & nbV( size_t i, size_t _I ) { return G.nb1(i)[_I]; }
        /// Provides read access to neighbor of factor
        const Neighbor & nbF( size_t I, size_t _i ) const { return G.nb2(I)[_i]; }
        /// Provides full access to neighbor of factor
        Neighbor & nbF( size_t I, size_t _i ) { return G.nb2(I)[_i]; }

        /// Returns the index of a particular variable
        size_t findVar( const Var & n ) const {
            size_t i = find( vars().begin(), vars().end(), n ) - vars().begin();
            assert( i != nrVars() );
            return i;
        }

        /// Returns a set of indexes corresponding to a set of variables
        std::set<size_t> findVars( VarSet &ns ) const {
            std::set<size_t> indexes;
            for( VarSet::const_iterator n = ns.begin(); n != ns.end(); n++ )
                indexes.insert( findVar( *n ) );
            return indexes;
        }

        /// Returns index of the first factor that depends on the variables
        size_t findFactor(const VarSet &ns) const {
            size_t I;
            for( I = 0; I < nrFactors(); I++ )
                if( factor(I).vars() == ns )
                    break;
            assert( I != nrFactors() );
            return I;
        }

        /// Return all variables that occur in a factor involving the i'th variable, itself included
        VarSet Delta( unsigned i ) const;

        /// Return all variables that occur in a factor involving some variable in ns, ns itself included
        VarSet Delta( const VarSet &ns ) const;

        /// Return all variables that occur in a factor involving the i'th variable, n itself excluded
        VarSet delta( unsigned i ) const;

        /// Return all variables that occur in a factor involving some variable in ns, ns itself excluded
        VarSet delta( const VarSet & ns ) const {
            return Delta( ns ) / ns;
        }

        /// Set the content of the I'th factor and make a backup of its old content if backup == true
        virtual void setFactor( size_t I, const Factor &newFactor, bool backup = false ) {
            assert( newFactor.vars() == factor(I).vars() ); 
            if( backup )
                backupFactor( I );
            _factors[I] = newFactor; 
        }

        /// Set the contents of all factors as specified by facs and make a backup of the old contents if backup == true
        virtual void setFactors( const std::map<size_t, Factor> & facs, bool backup = false ) {
            for( std::map<size_t, Factor>::const_iterator fac = facs.begin(); fac != facs.end(); fac++ ) {
                if( backup )
                    backupFactor( fac->first );
                setFactor( fac->first, fac->second );
            }
        }

        /// Clamp variable n to value i (i.e. multiply with a Kronecker delta \f$\delta_{x_n, i}\f$);
        /// If backup == true, make a backup of all factors that are changed
        virtual void clamp( const Var & n, size_t i, bool backup = false );

        /// Set all factors interacting with the i'th variable 1
        virtual void makeCavity( unsigned i, bool backup = false );

        /// Backup the factors specified by indices in facs
        virtual void backupFactors( const std::set<size_t> & facs );

        /// Restore all factors to the backup copies
        virtual void restoreFactors();

        /// Returns true if the FactorGraph is connected
        bool isConnected() const { return G.isConnected(); }

        /// Returns true if the FactorGraph is a tree
        bool isTree() const { return G.isTree(); }

        /// Returns true if each factor depends on at most two variables
        bool isPairwise() const;

        /// Returns true if each variable has only two possible values
        bool isBinary() const;

        /// Reads a FactorGraph from a file
        void ReadFromFile(const char *filename);

        /// Writes a FactorGraph to a file
        void WriteToFile(const char *filename) const;

        /// Writes a FactorGraph to a GraphViz .dot file
        void printDot( std::ostream& os ) const;
        
        /// Returns the cliques in this FactorGraph
        std::vector<VarSet> Cliques() const;

        /// Clamp variable v_i to value state (i.e. multiply with a Kronecker delta \f$\delta_{x_{v_i},x}\f$);
        /** This version changes the factor graph structure and thus returns a newly constructed FactorGraph
         *  and keeps the current one constant, contrary to clamp()
         */
        FactorGraph clamped( const Var & v_i, size_t x ) const;

        /// Returns a copy of *this, where all factors that are subsumed by some larger factor are merged with the larger factors.
        FactorGraph maximalFactors() const;

        /// Makes a backup of the I'th Factor
        void restoreFactor( size_t I );

        /// Restores the I'th Factor from the backup (it should be backed up first)
        void backupFactor( size_t I );

        /// Makes a backup of all factors connected to a set of variables
        void backupFactors( const VarSet &ns );
        /// Restores all factors connected to a set of variables from their backups
        void restoreFactors( const VarSet &ns );

        // Friends
        friend std::ostream& operator << (std::ostream& os, const FactorGraph& fg);
        friend std::istream& operator >> (std::istream& is, FactorGraph& fg);

        size_t VV2E(size_t n1, size_t n2) const { return G.VV2E(n1,n2); }
        const Edge& edge(size_t e) const { return G.edge(e); }
        void indexEdges() { G.indexEdges(); }
        size_t nr_edges() const { return G.nr_edges(); }
        const std::vector<Edge>& edges() const { return G.edges(); }

    private:
        /// Part of constructors (creates edges, neighbors and adjacency matrix)
        void constructGraph( size_t nrEdges );
};


template<typename FactorInputIterator, typename VarInputIterator>
FactorGraph::FactorGraph(FactorInputIterator fact_begin, FactorInputIterator fact_end, VarInputIterator var_begin, VarInputIterator var_end, size_t nr_fact_hint, size_t nr_var_hint ) : G(), _backup() {
    // add factors
    size_t nrEdges = 0;
    _factors.reserve( nr_fact_hint );
    for( FactorInputIterator p2 = fact_begin; p2 != fact_end; ++p2 ) {
        _factors.push_back( *p2 );
        nrEdges += p2->vars().size();
    }

    // add variables
    _vars.reserve( nr_var_hint );
    for( VarInputIterator p1 = var_begin; p1 != var_end; ++p1 )
        _vars.push_back( *p1 );

    // create graph structure
    constructGraph( nrEdges );
}


} // end of namespace dai


#endif
