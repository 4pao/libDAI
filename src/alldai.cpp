/*  Copyright (C) 2006-2008  Joris Mooij  [joris dot mooij at tuebingen dot mpg dot de]
    Radboud University Nijmegen, The Netherlands /
    Max Planck Institute for Biological Cybernetics, Germany

    This file is part of libDAI.

    libDAI is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    libDAI is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with libDAI; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/


#include <string>
#include <dai/alldai.h>
#include <dai/properties.h>
#include <dai/exceptions.h>


namespace dai {


using namespace std;


InfAlg *newInfAlg( const string &name, const FactorGraph &fg, const PropertySet &opts ) {
    if( name == ExactInf::Name )
        return new ExactInf (fg, opts);
#ifdef DAI_WITH_BP
    if( name == BP::Name ) 
        return new BP (fg, opts);
#endif
#ifdef DAI_WITH_MF
    if( name == MF::Name ) 
        return new MF (fg, opts);
#endif
#ifdef DAI_WITH_HAK
    if( name == HAK::Name ) 
        return new HAK (fg, opts);
#endif
#ifdef DAI_WITH_LC
    if( name == LC::Name )
        return new LC (fg, opts);
#endif
#ifdef DAI_WITH_TREEEP
    if( name == TreeEP::Name )
        return new TreeEP (fg, opts);
#endif
#ifdef DAI_WITH_JTREE
    if( name == JTree::Name )
        return new JTree (fg, opts);
#endif
#ifdef DAI_WITH_MR
    if( name == MR::Name )
        return new MR (fg, opts);
#endif
#ifdef DAI_WITH_GIBBS
    if( name == Gibbs::Name )
        return new Gibbs (fg, opts);
#endif
#ifdef DAI_WITH_CBP
    if( name == CBP::Name )
        return new CBP (fg, opts);
#endif
    DAI_THROW(UNKNOWN_DAI_ALGORITHM);
}


/// \todo Make alias file non-testdai-specific, and use it in newInfAlgFromString
InfAlg *newInfAlgFromString( const std::string &s, const FactorGraph &fg ) {
    string::size_type pos = s.find_first_of('[');
    string name;
    PropertySet opts;
    if( pos == string::npos ) {
        name = s;
    } else {
        name = s.substr(0,pos);

        stringstream ss;
        ss << s.substr(pos,s.length());
        ss >> opts;
    }
    return newInfAlg(name,fg,opts);
}


} // end of namespace dai
