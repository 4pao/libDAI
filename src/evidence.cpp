/*  Copyright (C) 2009  Charles Vaske  [cvaske at soe dot ucsc dot edu]
    University of California Santa Cruz

    This file is part of libDAI.

    libDAI is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    libDAI is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with libDAI; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/


#include <sstream>
#include <string>
#include <cstdlib>

#include <dai/util.h>
#include <dai/evidence.h>


namespace dai {


void Observation::addObservation( Var node, size_t setting ) {
    _obs[node] = setting;
}


void Observation::applyEvidence( InfAlg &alg ) const {
    for( std::map<Var, size_t>::const_iterator i = _obs.begin(); i != _obs.end(); ++i )
        alg.clamp( alg.fg().findVar(i->first), i->second );
}
  

void Evidence::addEvidenceTabFile( std::istream &is, FactorGraph &fg ) {
    std::map<std::string, Var> varMap;
    for( std::vector<Var>::const_iterator v = fg.vars().begin(); v != fg.vars().end(); ++v ) {
        std::stringstream s;
        s << v->label();
        varMap[s.str()] = *v;
    }

    addEvidenceTabFile( is, varMap );
}


void Evidence::addEvidenceTabFile( std::istream &is, std::map<std::string, Var> &varMap ) {
    std::string line;
    getline( is, line );
    
    // Parse header
    std::vector<std::string> header_fields;
    tokenizeString( line, header_fields );
    std::vector<std::string>::const_iterator p_field = header_fields.begin();
    if( p_field == header_fields.end() ) 
        DAI_THROW(INVALID_EVIDENCE_FILE);

    std::vector<Var> vars;
    for( ; p_field != header_fields.end(); ++p_field ) {
        std::map<std::string, Var>::iterator elem = varMap.find( *p_field );
        if( elem == varMap.end() )
            DAI_THROW(INVALID_EVIDENCE_FILE);
        vars.push_back( elem->second );
    }
    
    // Read samples
    while( getline(is, line) ) {
        std::vector<std::string> fields;
        tokenizeString( line, fields );
        if( fields.size() != vars.size() ) 
            DAI_THROW(INVALID_EVIDENCE_FILE);
        
        Observation sampleData;
        for( size_t i = 0; i < vars.size(); ++i ) {
            if( fields[i].size() > 0 ) { // skip if missing observation
                if( fields[i].find_first_not_of("0123456789") != std::string::npos )
                    DAI_THROW(INVALID_EVIDENCE_FILE);
                size_t state = atoi( fields[i].c_str() );
                if( state >= vars[i].states() )
                    DAI_THROW(INVALID_EVIDENCE_FILE);
                sampleData.addObservation( vars[i], state );
            }
        }
        _samples.push_back( sampleData );
    } // finished sample line
}


} // end of namespace dai
