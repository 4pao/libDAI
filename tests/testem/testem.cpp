/*  Copyright (C) 2009  Charles Vaske  [cvaske at soe dot ucsc dot edu]
    University of California Santa Cruz

    This file is part of libDAI.

    libDAI is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    libDAI is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with libDAI; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/


#include <iostream>
#include <fstream>
#include <string>

#include <dai/factorgraph.h>
#include <dai/evidence.h>
#include <dai/alldai.h>


using namespace std;
using namespace dai;


void usage( const string &msg ) {
    cerr << msg << endl;
    cerr << "Usage:" << endl;
    cerr << " testem factorgraph.fg evidence.tab emconfig.em" << endl;
    exit( 1 );
}


int main( int argc, char** argv ) {
    if( argc != 4 )
        usage("Incorrect number of arguments.");
    
    FactorGraph fg;
    fg.ReadFromFile( argv[1] );

    PropertySet infprops;
    infprops.Set( "verbose", (size_t)1 );
    infprops.Set( "updates", string("HUGIN") );
    InfAlg* inf = newInfAlg( "JTREE", fg, infprops );
    inf->init();

    Evidence e;
    ifstream estream( argv[2] );
    e.addEvidenceTabFile( estream, fg );

    cout << "Number of samples: " << e.nrSamples() << endl;
    for( Evidence::iterator ps = e.begin(); ps != e.end(); ps++ )
        cout << "Sample #" << (ps - e.begin()) << " has " << ps->observations().size() << " observations." << endl;

    ifstream emstream( argv[3] );
    EMAlg em(e, *inf, emstream);

    while( !em.hasSatisfiedTermConditions() ) {
        Real l = em.iterate();
        cout << "Iteration " << em.getCurrentIters() << " likelihood: " << l <<endl;
    }

    cout << endl << "Inferred Factor Graph:" << endl << "######################" << endl;
    cout.precision(12);
    cout << inf->fg();

    delete inf;

    return 0;
}
