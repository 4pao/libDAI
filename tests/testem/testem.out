Number of samples: 20
Sample #0 has 2 observations.
Sample #1 has 2 observations.
Sample #2 has 2 observations.
Sample #3 has 2 observations.
Sample #4 has 2 observations.
Sample #5 has 2 observations.
Sample #6 has 2 observations.
Sample #7 has 2 observations.
Sample #8 has 2 observations.
Sample #9 has 2 observations.
Sample #10 has 2 observations.
Sample #11 has 2 observations.
Sample #12 has 2 observations.
Sample #13 has 2 observations.
Sample #14 has 2 observations.
Sample #15 has 2 observations.
Sample #16 has 2 observations.
Sample #17 has 2 observations.
Sample #18 has 2 observations.
Sample #19 has 2 observations.
Iteration 1 likelihood: -13.8629
Iteration 2 likelihood: -9.56675
Iteration 3 likelihood: -9.56675

Inferred Factor Graph:
######################
1

2
0 1 
2 2 
4
0   0.166666666667
1   0.666666666667
2   0.833333333333
3   0.333333333333
Number of samples: 20
Sample #0 has 2 observations.
Sample #1 has 2 observations.
Sample #2 has 2 observations.
Sample #3 has 2 observations.
Sample #4 has 2 observations.
Sample #5 has 2 observations.
Sample #6 has 2 observations.
Sample #7 has 2 observations.
Sample #8 has 2 observations.
Sample #9 has 2 observations.
Sample #10 has 2 observations.
Sample #11 has 2 observations.
Sample #12 has 2 observations.
Sample #13 has 2 observations.
Sample #14 has 2 observations.
Sample #15 has 2 observations.
Sample #16 has 2 observations.
Sample #17 has 2 observations.
Sample #18 has 2 observations.
Sample #19 has 2 observations.
Iteration 1 likelihood: 0
Iteration 2 likelihood: 3.97035
Iteration 3 likelihood: 3.97035

Inferred Factor Graph:
######################
1

3
0 1 2 
2 2 2 
8
0   0.214285714286
1   0.642857142857
2   0.785714285714
3   0.357142857143
4   0.214285714286
5   0.642857142857
6   0.785714285714
7   0.357142857143
Number of samples: 5
Sample #0 has 5 observations.
Sample #1 has 4 observations.
Sample #2 has 6 observations.
Sample #3 has 6 observations.
Sample #4 has 5 observations.
Iteration 1 likelihood: 11.1646
Iteration 2 likelihood: 1.53723
Iteration 3 likelihood: 1.64691
Iteration 4 likelihood: 1.67497
Iteration 5 likelihood: 1.68191

Inferred Factor Graph:
######################
3

3
2 6 7 
2 2 2 
8
0   0.398343360806
1   0.351464146565
2   0.601656639194
3   0.648535853435
4   0.804844687957
5    0.67374245803
6   0.195155312043
7    0.32625754197

3
0 1 6 
2 2 2 
8
0    1.03521336269
1    1.54747895212
2    2.31765218974
3    1.28041900719
4      4.922079813
5    2.52725575019
6   0.831279296316
7   0.262805630803

3
1 2 4 
2 2 2 
8
0   0.398343360806
1   0.601656639194
2   0.351464146565
3   0.648535853435
4   0.804844687957
5   0.195155312043
6    0.67374245803
7    0.32625754197
Number of samples: 5
Sample #0 has 5 observations.
Sample #1 has 4 observations.
Sample #2 has 6 observations.
Sample #3 has 6 observations.
Sample #4 has 5 observations.
Iteration 1 likelihood: 11.1646
Iteration 2 likelihood: -7.29331
Iteration 3 likelihood: -7.261

Inferred Factor Graph:
######################
3

3
2 6 7 
2 2 2 
8
0   0.495312199726
1   0.499107942908
2   0.504687800274
3   0.500892057092
4   0.640416549958
5   0.495122293924
6   0.359583450042
7   0.504877706076

3
0 1 6 
2 2 2 
8
0   0.495312199726
1   0.504687800274
2   0.499107942908
3   0.500892057092
4   0.640416549958
5   0.359583450042
6   0.495122293924
7   0.504877706076

3
1 2 4 
2 2 2 
8
0   0.495312199726
1   0.504687800274
2   0.499107942908
3   0.500892057092
4   0.640416549958
5   0.359583450042
6   0.495122293924
7   0.504877706076
